docs/proposals/source-verificiation-policies.md

```diff
@@ -0,0 +1,313 @@
+* `trustedSigners` is the list of allowed signers
+
+### Anatomy of a source verification policy
+
+A source verification policy describes requirements for a source, as used by an Argo CD Application, that have to be met in order to be used for a sync.
+
+A source verification policy has the following properties:
+
+* A repository pattern (shell-glob type) that is matched against the URL of the source (mandatory)
+* A repository type that is matched against the application's repository type (mandatory)
+* A verification level that describes with which level the verification should be performed (mandatory)
+* A verification methods that describes how to the source should be verified (mandatory)
+* A list of identities of trusted signers (optional, if omitted, any known signer is trusted)
+
+Additionally, there are some optional properties you can set for a policy that depend on the verification level, method and respository type.
+
+Source verification policies are configured within an AppProject and should be considered an admin-level configurable.
+
+#### Repository pattern
+
+This is a glob-style pattern that is matched against the URL of the source to verify. When the pattern matches a source, it will be verified, otherwise verification of the source will be skipped.
+
+#### Repository type
+
+The type of repository. Right now, only "git" is supported.
+
+#### Verification levels
+
+Each source verification policy currently knows about four levels of verification:
+
+* `none`: As the name implies, verification will be disabled at this verification level.
+* `head`: This verification level will verify the commit that is pointed to by the HEAD of the target revision. In case target revision is an annotated tag, only the signature of the tag will be verified.
+* `progressive`: This verification level will verify all commits between the application's target revision and the revision it was last synced to, but not beyond that. The revision that the application is currently synced to does not necessarily need to be signed, so that a migration path from level `head` to the more secure method `progressive` is possible. However, if the application hasn't synced before (i.e. there is no previous revision in the application's sync history), the `progressive` level will inhibit the same behaviour as the `strict` verification level (see below).
+* `strict`: This verification level will always verify all commits in the history of the repository that led to the application's target revision. The `strict` level effectively enforces that all commits in your repository need to be signed from the beginning of your repository's history, without allowing _any_ unsigned commit in the history. This is the most secure policy.
+
+For the `strict` and `progressive` verification levels, if the target revision is an annotated tag, then that tag's signature will also be verified in addition to the specific commits verified as defined by the respective level.
```

Comment:
Currently either a signed tag or a lightweight tag are accepted. In these modes, should only signed tags be allowed? A malicious user, or git server, could move lightweight tags to an older commit to perform a downgrade attack (e.g. deploy and older version of an app with known vulnerabilities)

Reply:
Hm, that's a good catch. I haven't thought much about this scenario. However, there will be another attack vector here, which is the Application CR. The same downgrade attack would be possible by just modifying the source in the Application to resolve a previous tag. But yeah, I agree, that this would add up to a defense-in-depth principle.

What do you think about making this another option for a policy? For example, `enforceSignedTag: <bool>` or `allowSignedTagsOnly: <bool>`?

Reply:
Yes, I think an additional option to force tags to be signed would solve the issue.

I was worried about something going bad in git - I think if a malicious user has access to the Application CR then it's probably game over anyway?

Reply:
OK, basically, someone with write access to the Application CR can modify the targetRevision to sync against. 

The scope of the verification policies is to only allow sync against previously vetted (e.g. signed) revisions. To my knowledge, neither PGP nor Git knows about the concept of signature revocation, except when revoking the private key used to make the signature. I think preventing downgrade attacks becomes rather futile on this level.

But let's keep thinking about it.

Reply:
My concern was not really signature revocation, it was more what mischief a malicious git server could achieve. When the Application is tracking a branch it should only accept revisions that are descendants of the currently synced revision and if it's tracking a tag it should be a signed tag, both to prevent the git server moving back to a previous (but signed) vulnerable commit. If a user has the ability to change the Application CR, then they can legitimately point targetRevision where they like, as long as it's signed